Cloudkit AAP config-as-code (#23)
innabox/cloudkit-aap: collections/ansible_collections/cloudkit/config_as_code/README.md

# Ansible Collection - cloudkit.config_as_code

This role defines the configuration as code of AAP to run the Ansible playbooks
that are part of CloudKit.

## Credentials

> Since AAP runs in an Openshift cluster, we expect the required credentials to

Reviewer: As long as we're capitalizing this word, we may as well get the fully correct "OpenShift" here and throughout the PR.

Author: done!

be passed as environment variables. These environment variables are defined as
`Secrets` in OpenShift, and used by a container group.

We define 2 container groups in AAP, one used to reconcile the configuration of
AAP itself, and one used for cluster fulfillment operations.

### Config-as-code environment variables

In order to reconcile the configuration of AAP, we require the following
variables:

- `RH_USERNAME` and `RH_PASSWORD`: your Red Hat console username and password,
  used to manage the subscription of AAP
- `AAP_HOSTNAME`: URL of the AAP instance to be configured
- `AAP_VALIDATE_CERTS`: true if the SSL certificate behind `AAP_HOSTNAME`
  should be checked
- `AAP_ORGANIZATION_NAME`: the AAP organization that should be created
- `AAP_PROJECT_NAME`: the name of the project to be created
- `AAP_PROJECT_GIT_URI`: the repository that is behind the AAP project
- `AAP_PROJECT_GIT_BRANCH`: the git branch to use
- `AAP_EE_IMAGE`: the registry URL to the execution environment image

These variables must be defined in a secret named:

> `${AAP_ORGANIZATION_NAME}-${AAP_PROJECT_NAME}-config-as-code-ig` in the

Reviewer 2: What does the `-ig` suffix mean?

Reviewer 2: ...ah, I guess that is "instance group".

Author: yep, I added that suffix, not sure if it makes sense for a secret name

namespace where AAP is deployed.

Note: since the container group is configured to run in the same namespace as
the AAP instance, the admin credentials to log against AAP are injected into
the pod.

### Cluster fulfillment environment variables

Here we need to define all the credentials required by the cluster fulfillment
use case, e.g.:

- `AWS_*`: AWS credentials
- `OS_*`: OpenStack credentials
- ...whatever in needed

These variables must be defined in a secret named:
`${AAP_ORGANIZATION_NAME}-${AAP_PROJECT_NAME}-cluster-fulfillment-ig` in the
namespace where AAP is deployed.

The cluster fulfillment needs to access the Kube API of the cluster it runs on,
so we expect a service account `cloudkit-sa` to exists with enough rights (TBD).

## Deploy a local AAP installation using CRC

### Download and deploy CRC

> CRC basically OpenShift in a VM, you can download it from Red Hat Console:

Reviewer 2: ```suggestion
CRC is basically OpenShift in a VM, you can download it from Red Hat Console:
```

https://console.redhat.com/openshift/downloads. You will need `oc` which is
downloadable from the same place.

Extract the package, put the binary in your path, start the cluster, and log as
`kubeadmin`:

```
crc start
oc login -u kubeadmin  https://api.crc.testing:6443
```

> ### Deploy AAP

Reviewer 2: There seems to be a step missing here; you've got (1) set up a CRC environment then (2) create an AnsibleAutomationPlatform CR, but where is the "install AAP" step? Or does CRC now include AAP out of the box?

Author: Right, I forgot to talk about the subscription, I will add it, thanks!

Author: Added!

> Create a new namespace and deploy an AAP instance using the Ansible:

Reviewer 2: ```suggestion
Create a new namespace and deploy an AAP instance using these manifests:
```

  AAP_PROJECT_GIT_URI: <the git repository where your playbooks and rulebooks live>
  AAP_PROJECT_GIT_BRANCH: <the git branch to be tracked>
  AAP_EE_IMAGE: <the execution environment image>
---
apiVersion: v1
kind: Secret
metadata:
  name: <your AAP organization>-<your AAP project>-cluster-fulfillment-ig
type: Opaque
stringData:
  <setup here the credentials required for cluster fulfillment playbooks (AWS, OpenStack, ...)>
EOF

oc apply -f cloudkit_env.yml -n fulfillment-aap
```

#### Option 1: from your local environment

Set your environment (same as above in `<your AAP organization>-<your AAP
project>-config-as-code-ig` secret):

```
export RH_USERNAME=...
export RH_PASSWORD=...
export AAP_HOSTNAME=...
export AAP_USERNAME=...
export AAP_PASSWORD=...
export AAP_VALIDATE_CERTS=...
export AAP_ORGANIZATION_NAME=...
export AAP_PROJECT_NAME=...
export AAP_PROJECT_GIT_URI=...
export AAP_PROJECT_GIT_BRANCH=...
export AAP_EE_IMAGE=...
```

Run subscription and configure playbooks:
```
ansible-playbook cloudkit.config_as_code.subscription
ansible-playbook cloudkit.config_as_code.configure
```

#### Option 2: as a job running in Kubernetes

> If this collection is backed in a container image (e.g.: in the custom

Reviewer 2: Do you mean "baked" here? I would probably rephrase this:

```suggestion
If this collection is being used in a container image (e.g.: in the custom
```

execution environment used in AAP), there is the possibility to run the
configuration in a k8s job:

```
cat << EOF > job.yml
apiVersion: batch/v1
kind: Job
metadata:
  name: aap-bootstrap
spec:
  template:
    spec:
      containers:
        - image: <your container image with cloudkit.config_as_code collection in it>

>           name: boostrap

Reviewer 2: ```suggestion
          name: bootstrap
```

          args:
            - ansible-playbook
            - "cloudkit.config_as_code.subscription"

>             - "cloudkit.config_as_code.configure"

Reviewer 2: Is there a particular version of Ansible required to make this work? Trying to run these playbooks using their fully qualified names with Ansible 2.18.5 from the `cloudkit-aap` directory results in:

```
ERROR! the playbook: cloudkit.config_as_code.subscription could not be found
```

Author: You can try `ANSIBLE_COLLECTIONS_PATHS=collections/ ansible-playbook cloudkit.config_as_code.subscription` . The container image is configured to look for the collection at the right place.

Author: I added the `collection_paths` setting to the `ansible.cfg`
